File: leetcode/91/P91.rb
# 91. Decode Ways
# @param {String} s
# @return {Integer}
def num_decodings(s)
	n = s.size
	return 0 if n == 0
	d = Array.new n + 1, 0
	d[-1] = 1
	for i in 0...n
		if s[i] != '0'
			d[i] = d[i - 1]
		end
		num = s[i - 1..i].to_i
		if num >= 10 && num <= 26
			d[i] += d[i - 2]
		end

	end
	return d[n - 1]
end

p num_decodings '12'

File: leetcode/91/P91.java
public class Solution {
  public int numDecodings(String s) {
    int n = s.length();
    if (n == 0) return 0;
    int[] d = new int[n];
    for (int i = 0; i < n; i++) {
      int num = s.charAt(i) - '0';
      if (num > 0) d[i] = i == 0 ? 1 : d[i - 1];
      if (i > 0) {
        int lastNum = s.charAt(i - 1) - '0';
        int val = lastNum * 10 + num;
        if (val >= 10 && val <= 26)
          d[i] += i - 2 >= 0 ? d[i - 2] : 1;
      }
    }
    return d[n - 1];
  }
}